pkg/lib/form-layout.less

```diff
@@ -0,0 +1,247 @@
+:root {
+  // CSS variable to define the number of (label + control) columns.
+  // It dynamically changes on narrow pages (see media query below).
+  --ct-form-layout-columns: 2;
+}
+
+// Cockpit Form Layout: Automatically have Cockpit display your form in
+// an optimal layout.
+//
+// By default, all labels are aligned and sized properly and form elements
+// stretch to take up the remaining space.
+//
+//
+// There are additional classes and attributes you can add to each
+// control directly under `ct-form-layout`:
+//
+// `ct-form-layout-split`: The grid can be split on a `form-control`
+// level by adding a this class. If you want two elements next to each
+// other, both should have this class.
+//
+// `ct-form-layout-relax`: Form elements normally stretch to take up the
+// full space. You can relax their width by adding this class to the
+// control. Inputs with a size attribute are auto-relaxed and do not
+// need this class.
+//
+// `ct-form-layout-stretch`: If a control has a width specified
+// elsewhere, you can force it to stretch. This is mainly useful when
+// using <div role="group"> to group elements.
+//
+// role="group": When there are two related elements, such as a text
+// input and a dropdown, you can group them together using this HTML
+// attribute. It's similar in purpose to a <fieldset>, but works for
+// layouts in Chrome (unlike fieldset). This can be attached to any
+// container element, but will most likely be used with <div>. The role
+// adds semantic meaning to the element for screen readers, and we key
+// the CSS off of the role.
+//
+// <hr>: While this is an element, it has a special meaning and is used
+// to add some vertical spacing to a form.
+//
+// Most of the time, you can ignore all the optional classes (and
+// attribute and hr element) and simply wrap your labels & controls in
+// a <form class="ct-form-layout"> and layout magic happens.
+
+.ct-form-layout {
+  display: grid;
+  grid-gap: 0 10px;
+  // Repeat a label that is a minimum of 4em and its control that
+  // fills the remaining space by a CSS variable (default: 2)
+  grid-template-columns: repeat(var(--ct-form-layout-columns), minmax(4em, max-content) 1fr);
+  place-items: baseline stretch;
+
+  // Horizontal rules directly under a form-layout container serve to
+  // add some vertical space in forms. This is useful for visually
+  // grouping similar elements with whitespace.
+  //
+  // It's not the same as actually grouping elements (which can be done
+  // in the usual ways as well as adding a role="group".
+  > hr {
+    border: none;
+    // LESS needs this to be escaped with ~"". You'll see it below too.
+    // CSS wants the string to be 1 / -1 without escaping.
+    grid-column: ~"1 / -1";
+    height: 15px;
+    margin: 0;
+    padding: 0;
+  }
+
+  label {
+    // Make labels are the same height as controls,
+    // to ensure vertical alignment happens properly
+    line-height: 26px;
+  }
```

Comment:
This looks a bit dubious - what happens if the font size is bigger? (HiDPI or zooming)

Reply:
A CSS px is not a pixel, even though on a non-HiDPI screen it is rendered as a pxiel. It may display differently on other devices and also on printouts. It also takes into account HiDPI and zooming.

https://www.w3.org/Style/Examples/007/units.en.html#units

Generally, I'm not a fan of px for sizing fonts, but I'm using what PatternFly does. I meant to replace some of the magic numbers (including this) with PatternFly variables in LESS. I'll look into doing this today. The end result should be identical, but the values will then be self-explanatory and change if it ever changes in PatternFly.

Reply:
26px is font-size + padding + border. I looked at the size of the buttons, dropdowns, etc. and saw they're all at 26px, so I fixed the height for elements to have that same line height within our forms. It changes a few elements by a few pixels, but then things are no longer off by a few pixels vertically. :wink:

This is one of the reasons why I decided calling it form-layout instead of form-grid or grid-forms — it's doing more than just CSS grid stuff. Plus, it probably shouldn't be named on the tech used, but by what it does.

Reply:
Note: As it wouldn't change anything in the resulting CSS, I could change the values to their semantic meanings _after_ this PR lands, in a follow-up.